### PR TITLE
Dao curry twice

### PR DESCRIPTION
**Pull request by matt-o-how**
### Purpose:
Optimise the lockup and proposal chialisp files to use the curry twice technique, making the cost of calculating their puzzle hashes dramatically lower elsewhere in code.

Also fixed a couple of other minor bugs while I was in there.

### CODE REVIEW

File: tests/wallet/dao_wallet/test_dao_wallets.py [hunk 1]
```diff
@@ -581,6 +581,7 @@ async def test_dao_proposals(self_hostname: str, three_wallet_nodes: SimulatorsA
     await asyncio.sleep(1)
     for i in range(1, num_blocks):
         await full_node_api.farm_new_transaction_block(FarmNewBlockProtocol(puzzle_hash_0))
+        await asyncio.sleep(2)
```
Comment (geoffwalmsley) [inline tests/wallet/dao_wallet/test_dao_wallets.py:584]: I usually add the sleep after the for loop exits, it doesn't need to sleep after each block. Unless you found this is the only way it passes?

File: chia/wallet/puzzles/dao_proposal_validator.clsp [hunk 2]
```diff
@@ -42,70 +36,73 @@
...
         )
-        coin_amount
-      )
-    )
-    (c
-      (list CREATE_COIN PROPOSAL_EXCESS_PAYOUT_PUZ_HASH (- coin_amount 1))
-      (c
-        (list CREATE_PUZZLE_ANNOUNCEMENT proposal_id)  ; specify the proposal we're talking about
-        conditions
+        (x "three")
```
Comment (geoffwalmsley) [inline chia/wallet/puzzles/dao_proposal_validator.clsp:101]: Can probably remove the error messages

File: chia/wallet/puzzles/dao_proposal_validator.clsp [hunk 3]
```diff
@@ -42,70 +36,73 @@
...
-            PROPOSAL_MOD_HASH
-            (c (f SINGLETON_STRUCT) (c proposal_id (r (r SINGLETON_STRUCT))))
-            PROPOSAL_TIMER_MOD_HASH
-            CAT_MOD_HASH
-            DAO_FINISHED_STATE_MOD_HASH
-            TREASURY_MOD_HASH
-            LOCKUP_MOD_HASH
-            CAT_TAIL_HASH
-            (f (r SINGLETON_STRUCT))
+    (if
```
Comment (geoffwalmsley) [inline chia/wallet/puzzles/dao_proposal_validator.clsp:57]: how come we use nested if's instead of an assert now?

File: chia/wallet/cat_wallet/dao_cat_wallet.py [hunk 4]
```diff
@@ -387,42 +384,24 @@ async def create_vote_spend(
...
-                        INNERPUZHASH,
-                    ],
+                    proposal_puzzle.get_tree_hash(),
                     vote_info,
                     vote_amount,
                     lci.inner_puzzle.get_tree_hash(),
                     0,
                 ]
             )
-            # breakpoint()
```
Comment (Quexington) [inline chia/wallet/cat_wallet/dao_cat_wallet.py:425]: Good comment

File: chia/wallet/puzzles/dao_proposal.clsp [hunk 5]
```diff
@@ -398,14 +342,10 @@
...
-                      PROPOSAL_MOD_HASH
                       PROPOSAL_TIMER_MOD_HASH
-                      LOCKUP_MOD_HASH
-                      DAO_FINISHED_STATE_MOD_HASH
-                      CAT_MOD_HASH
-                      CAT_TAIL_HASH
-                      SINGLETON_STRUCT
-                      TREASURY_ID
+                      SELF_HASH
+                      ; SINGLETON_STRUCT
```
Comment (geoffwalmsley) [inline chia/wallet/puzzles/dao_proposal.clsp:347]: can remove this comment

File: chia/wallet/puzzles/dao_proposal.clsp [hunk 6]
```diff
@@ -266,14 +216,14 @@
   (if self_destruct_flag
       ; assert self_destruct_time > proposal_timelock_length
       ; this is the code path for if we've not been accepted by the treasury for a long time, and we're "bad" for some reason
-      (assert (> self_destruct_time vote_coin_ids_or_proposal_timelock_length)
+      (if (> self_destruct_time vote_coin_ids_or_proposal_timelock_length)
         (list
-          (list CREATE_COIN (calculate_finished_state SINGLETON_STRUCT DAO_FINISHED_STATE_MOD_HASH) ONE (list TREASURY_ID))
+          (list CREATE_COIN (calculate_finished_state (c SINGLETON_MOD_HASH (c SINGLETON_ID LAUNCHER_PUZZLE_HASH)) DAO_FINISHED_STATE_MOD_HASH) ONE (list TREASURY_ID))
```
Comment (geoffwalmsley) [inline chia/wallet/puzzles/dao_proposal.clsp:221]: this would be a good place to use `let` to create singleton_struct so we don't construct it twice. Not a showstopper though.
Probably also applies to some of the announcements in this code block

File: chia/wallet/puzzles/dao_proposal_validator.clsp (hunk 2 above)
Comment (matt-o-how) [reply chia/wallet/puzzles/dao_proposal_validator.clsp:101]: good catch

File: chia/wallet/puzzles/dao_proposal_validator.clsp (hunk 3 above)
Comment (matt-o-how) [reply chia/wallet/puzzles/dao_proposal_validator.clsp:57]: for the sake of adding debug values to the (x) will fix

File: chia/wallet/puzzles/dao_proposal.clsp [hunk 7]
```diff
@@ -183,7 +148,7 @@
       my_amount
     )
     (if coin_id_list
-        (assert (> (f vote_amount_list) 0)
+        (if (> (f vote_amount_list) 0)
```
Comment (Quexington) [inline chia/wallet/puzzles/dao_proposal.clsp:151]: Why this change?  I think this is what assert does anyways yeah?

File: chia/wallet/puzzles/dao_proposal_validator.clsp [hunk 8]
```diff
@@ -42,70 +36,73 @@
...
-      (sha256 ONE PROPOSAL_MOD_HASH)
-      (sha256tree PROPOSAL_SINGLETON_STRUCT)
+      (sha256 ONE proposal_innerpuz_hash)
+      (sha256 ONE proposal_singleton_id)
+      (sha256 ONE PROPOSAL_SELF_HASH)
     )
   )
 
-  (assert
+  (if
```
Comment (Quexington) [inline chia/wallet/puzzles/dao_proposal_validator.clsp]: Again, this is a big readability concern, and I think assert does this as a macro

File: chia/wallet/puzzles/dao_proposal.clsp (hunk 5 above)
Comment (matt-o-how) [reply chia/wallet/puzzles/dao_proposal.clsp:347]: I'll move it to the side for clarity

File: tests/wallet/dao_wallet/test_dao_wallets.py (hunk 1 above)
Comment (Quexington) [inline tests/wallet/dao_wallet/test_dao_wallets.py:584]: `.sleep` == bad, check out `full_node_api.wait_for_wallet_synced`

File: tests/wallet/dao_wallet/test_dao_wallets.py [hunk 9]
```diff
@@ -810,7 +812,7 @@ async def get_dao_rules(wallet):
 
     # Give the wallet nodes a second
     await asyncio.sleep(1)
-    for i in range(1, 12):
+    for i in range(1, 20):
```
Comment (Quexington) [inline tests/wallet/dao_wallet/test_dao_wallets.py:815]: This is A LOT of blocks, this test will likely be very slow.  Again, I think you should check out some of the methods on the full node simulator, they're useful for waiting to progress in a test.

File: chia/wallet/puzzles/dao_proposal_validator.clsp (hunk 8 above)
Comment (matt-o-how) [reply chia/wallet/puzzles/dao_proposal_validator.clsp]: leaving it as if was a mistake, heldover from debugging fixed now

File: chia/wallet/puzzles/dao_proposal.clsp (hunk 7 above)
Comment (matt-o-how) [reply chia/wallet/puzzles/dao_proposal.clsp:151]: leaving it as if was a mistake, heldover from debugging fixed now

File: tests/wallet/dao_wallet/test_dao_wallets.py (hunk 1 above)
Comment (matt-o-how) [reply tests/wallet/dao_wallet/test_dao_wallets.py:584]: useful to know thank you

File: tests/wallet/dao_wallet/test_dao_wallets.py (hunk 9 above)
Comment (matt-o-how) [reply tests/wallet/dao_wallet/test_dao_wallets.py:818]: brought it down now

**Review by geoffwalmsley — APPROVED** [review]

**Review by Quexington — APPROVED** [review]
